<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Working with KneadData ¬∑ BiobakeryUtils.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="http://docs.ecojulia.org/BiobakeryUtils.jl/stable/kneaddata/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.044/juliamono.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">BiobakeryUtils.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">BiobakeryUtils</a></li><li><a class="tocitem" href="../gettingstarted/">Getting Started</a></li><li class="is-active"><a class="tocitem" href>Working with KneadData</a><ul class="internal"><li><a class="tocitem" href="#Installation-and-setup"><span>Installation and setup</span></a></li><li><a class="tocitem" href="#Running-on-a-single-end-sequencing-data"><span>Running on a single-end sequencing data</span></a></li><li><a class="tocitem" href="#Running-on-paired-end-sequencing-data"><span>Running on paired-end sequencing data</span></a></li><li><a class="tocitem" href="#Changing-Defaults"><span>Changing Defaults</span></a></li><li><a class="tocitem" href="#API-Reference"><span>API Reference</span></a></li></ul></li><li><a class="tocitem" href="../metaphlan/">Working with MetaPhlAn</a></li><li><a class="tocitem" href="../humann/">Working with HUMAnN</a></li><li><a class="tocitem" href="../microbiome/">Microbiome.jl Docstrings</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Working with KneadData</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Working with KneadData</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/main/docs/src/kneaddata.md" title="Edit on GitHub"><span class="docs-icon fab">ÔÇõ</span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="kneaddata-tutorial"><a class="docs-heading-anchor" href="#kneaddata-tutorial">KneadData Tutorial with BiobakeryUtils.jl</a><a id="kneaddata-tutorial-1"></a><a class="docs-heading-anchor-permalink" href="#kneaddata-tutorial" title="Permalink"></a></h1><ul><li>üóíÔ∏è This tutorial is meant to be run in parallel with / mirror the <a href="https://github.com/biobakery/biobakery/wiki/kneaddata">official KneadData</a></li><li>‚ùìÔ∏è If you have questions about MetaPhlAn itself, please direct them to the <a href="https://forum.biobakery.org/c/Microbial-community-profiling/MetaPhlAn">bioBakery help forum</a></li><li>ü§î If you have questions about using the MetaPhlAn tools in julia, <a href="https://github.com/EcoJulia/BiobakeryUtils.jl/issues/new/choose">please open an issue</a>, or start a discussion over on <a href="https://github.com/EcoJulia/Microbiome.jl/discussions/new"><code>Microbiome.jl</code></a>!</li><li>üìî For a function / type reference, <a href="#Functions-and-Types">jump to the bottom</a></li></ul><h2 id="Installation-and-setup"><a class="docs-heading-anchor" href="#Installation-and-setup">Installation and setup</a><a id="Installation-and-setup-1"></a><a class="docs-heading-anchor-permalink" href="#Installation-and-setup" title="Permalink"></a></h2><p>If you haven&#39;t already, check out the <a href="../gettingstarted/#getting-started">&quot;Getting Started&quot;</a> page to install julia, create an environment,xd and install BiobakeryUtils.jl, and hook up or install the MetaPhlAn v3 command line tools.</p><p>This tutorial assumes:</p><ol><li>You are running julia v1.6 or greater</li><li>You have activated a julia Project that has <code>BiobakeryUtils.jl</code> installed</li><li>The <code>kneaddata</code> python package is installed, and accessible from your <code>PATH</code>.</li></ol><p>If any of those things aren&#39;t true, or you don&#39;t know if they&#39;re true, go back to <a href="../gettingstarted/#getting-started">&quot;Getting Started&quot;</a> to see if you skipped a step. If you&#39;re still confused, please ask (see 3rd bullet point at the top)!</p><h3 id="Contamination-databases"><a class="docs-heading-anchor" href="#Contamination-databases">Contamination databases</a><a id="Contamination-databases-1"></a><a class="docs-heading-anchor-permalink" href="#Contamination-databases" title="Permalink"></a></h3><p>By default, kneaddata will only trim reads based on quality scores. If you would also like to remove contaminating sequences (eg from human or mouse DNA reads), you&#39;ll need to download them.</p><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.kneaddata_database" href="#BiobakeryUtils.kneaddata_database"><code>BiobakeryUtils.kneaddata_database</code></a> ‚Äî <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">kneaddata_database(db, kind, path)</code></pre><p>See <code>kneaddata_database --help</code></p><pre><code class="language-julia hljs">kneaddata_database(&quot;human_genome&quot;, &quot;bowtie2&quot;, &quot;/some/database/dir/&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/47c40f26fb364a745d5e3c4ccb1b72387b1798cc/src/kneaddata.jl#L70-L78">source</a></section></article><p>To see what databases are available, you need to use the command line, <code>kneaddata_database --available</code>.</p><h3 id="Demo-files"><a class="docs-heading-anchor" href="#Demo-files">Demo files</a><a id="Demo-files-1"></a><a class="docs-heading-anchor-permalink" href="#Demo-files" title="Permalink"></a></h3><p>The demo files for the kneaddata tutorial can be found in this package&#39;s <code>test</code> folder, which you can find with</p><pre><code class="language-julia-repl hljs">julia&gt; demo = abspath(joinpath(dirname(Base.find_package(&quot;BiobakeryUtils&quot;)), &quot;..&quot;, &quot;test&quot;, &quot;files&quot;, &quot;kneaddata&quot;));

julia&gt; readdir(demo)
10-element Vector{String}:
 &quot;SE_extra.fastq&quot;
 &quot;demo_db.1.bt2&quot;
 &quot;demo_db.2.bt2&quot;
 &quot;demo_db.3.bt2&quot;
 &quot;demo_db.4.bt2&quot;
 &quot;demo_db.rev.1.bt2&quot;
 &quot;demo_db.rev.2.bt2&quot;
 &quot;seq1.fastq&quot;
 &quot;seq2.fastq&quot;
 &quot;singleEnd.fastq&quot;</code></pre><h2 id="Running-on-a-single-end-sequencing-data"><a class="docs-heading-anchor" href="#Running-on-a-single-end-sequencing-data">Running on a single-end sequencing data</a><a id="Running-on-a-single-end-sequencing-data-1"></a><a class="docs-heading-anchor-permalink" href="#Running-on-a-single-end-sequencing-data" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/kneaddata#single-end-reads">Official tutorial link</a></p><p>You can use the <code>kneaddata</code> commandline tool using the <a href="#BiobakeryUtils.kneaddata"><code>kneaddata()</code></a> function from BiobakeryUtils.jl</p><pre><code class="language-julia hljs">julia&gt; kneaddata(joinpath(demo, &quot;singleEnd.fastq&quot;), &quot;kneaddataOutputSingleEnd&quot;; reference_db=joinpath(demo, &quot;demo_db&quot;))
‚îå Info: Running command: kneaddata -i /home/kevin/.julia/dev/BiobakeryUtils/test/files/kneaddata/singleEnd.fastq -o
‚îÇ kneaddataOutputSingleEnd --trimmomatic /home/kevin/.julia/conda/3/envs/BiobakeryUtils/share/trimmomatic -db
‚îî /home/kevin/.julia/dev/BiobakeryUtils/test/files/kneaddata/demo_db
Reformatting file sequence identifiers ...

Initial number of reads ( /tmp/jl_JXPuAs/kneaddataOutputSingleEnd/reformatted_identifiersjlcp_ry6_singleEnd ): 16902.0
# ... etc</code></pre><h2 id="Running-on-paired-end-sequencing-data"><a class="docs-heading-anchor" href="#Running-on-paired-end-sequencing-data">Running on paired-end sequencing data</a><a id="Running-on-paired-end-sequencing-data-1"></a><a class="docs-heading-anchor-permalink" href="#Running-on-paired-end-sequencing-data" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/kneaddata#paired-end-reads">Official tutorial link</a></p><p>To run on paired end data, simply pass an array of file paths to the <code>input</code> argument.</p><pre><code class="language-julia hljs">julia&gt; kneaddata([joinpath(demo, &quot;seq1.fastq&quot;), joinpath(demo, &quot;seq2.fastq&quot;)],
                   &quot;kneaddataOutputPairedEnd&quot;; reference_db=joinpath(demo, &quot;demo_db&quot;))
‚îå Info: Running command: kneaddata -i /home/kevin/.julia/dev/BiobakeryUtils/test/files/kneaddata/seq1.fastq -i
‚îÇ /home/kevin/.julia/dev/BiobakeryUtils/test/files/kneaddata/seq2.fastq -o kneaddataOutputPairedEnd --trimmomatic
‚îÇ /home/kevin/.julia/conda/3/envs/BiobakeryUtils/share/trimmomatic -db
‚îî /home/kevin/.julia/dev/BiobakeryUtils/test/files/kneaddata/demo_db
Initial number of reads ( /home/kevin/.julia/dev/BiobakeryUtils/test/files/kneaddata/seq1.fastq ): 42473.0
Initial number of reads ( /home/kevin/.julia/dev/BiobakeryUtils/test/files/kneaddata/seq2.fastq ): 42473.0
Running Trimmomatic ...
Total reads after trimming ( /tmp/jl_JXPuAs/kneaddataOutputPairedEnd/seq1_kneaddata.trimmed.1.fastq ): 35341.0
Total reads after trimming ( /tmp/jl_JXPuAs/kneaddataOutputPairedEnd/seq1_kneaddata.trimmed.2.fastq ): 35341.0
Total reads after trimming ( /tmp/jl_JXPuAs/kneaddataOutputPairedEnd/seq1_kneaddata.trimmed.single.1.fastq ): 5385.0
Total reads after trimming ( /tmp/jl_JXPuAs/kneaddataOutputPairedEnd/seq1_kneaddata.trimmed.single.2.fastq ): 847.0</code></pre><h2 id="Changing-Defaults"><a class="docs-heading-anchor" href="#Changing-Defaults">Changing Defaults</a><a id="Changing-Defaults-1"></a><a class="docs-heading-anchor-permalink" href="#Changing-Defaults" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/kneaddata#changing-defaults">Official tutorial link</a></p><p>To use the default-altering options, pass them as key words to the <code>kneaddata()</code> function.</p><p>Eg, to set maximum memory utilization to 200 Mb, add <code>max_memory=&quot;200m&quot;</code> to the function call.</p><h2 id="API-Reference"><a class="docs-heading-anchor" href="#API-Reference">API Reference</a><a id="API-Reference-1"></a><a class="docs-heading-anchor-permalink" href="#API-Reference" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.kneaddata" href="#BiobakeryUtils.kneaddata"><code>BiobakeryUtils.kneaddata</code></a> ‚Äî <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">kneaddata(inputfile, outputfile; kwargs...)</code></pre><p>Run <code>kneaddata</code> command line tool on <code>inputfile</code>, creating <code>outputfile</code>. Requires <code>kneaddata</code> to be installed and accessible in the <code>PATH</code> (see <a href="@ref">Getting Started</a>).</p><p><code>kneaddata</code> options can be passed via keyword arguments. For example, if on the command line you would run:</p><pre><code class="language-sh hljs">$ kneaddata -i some.fastq.gz -o test --n 8 --bypass-trim</code></pre><p>using this function, you would write:</p><pre><code class="language-julia hljs">kneaddata(&quot;some.fastq.gz&quot;, &quot;test&quot;; n = 8, bypass_trim=true)</code></pre><p>To pass multiple databases, pass an array of paths to the <code>reference_db</code> argument</p><p>Conda installations of <code>trimmomatic</code> (a dependency of <code>kneaddata</code>) don&#39;t work properly out of the box. If you have installed <code>kneaddata</code> using commandline conda (instead of <code>Conda.jl</code>), use <code>trimmomatic = /path/to/trimmomatic</code>, where <code>/path/to/trimmomatic</code> is something like <code>/home/username/miniconda3/envs/biobakery3/share/trimmomatic</code>. If you used <code>BiobakeryUtils.install_deps()</code>, you don&#39;t need to worry about this.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/47c40f26fb364a745d5e3c4ccb1b72387b1798cc/src/kneaddata.jl#L1-L34">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../gettingstarted/">¬´ Getting Started</a><a class="docs-footer-nextpage" href="../metaphlan/">Working with MetaPhlAn ¬ª</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.19 on <span class="colophon-date" title="Friday 1 July 2022 22:09">Friday 1 July 2022</span>. Using Julia version 1.7.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
