<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Working with HUMAnN ¬∑ BiobakeryUtils.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">BiobakeryUtils.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">BiobakeryUtils</a></li><li><a class="tocitem" href="../general/">General Utilities</a></li><li><a class="tocitem" href="../gettingstarted/">Getting Started</a></li><li class="is-active"><a class="tocitem" href>Working with HUMAnN</a><ul class="internal"><li><a class="tocitem" href="#Installation-and-setup"><span>Installation and setup</span></a></li><li><a class="tocitem" href="#Running-HUMAnN"><span>Running HUMAnN</span></a></li><li><a class="tocitem" href="#Default-outputs"><span>Default outputs</span></a></li><li><a class="tocitem" href="#Manipulating-tables"><span>Manipulating tables</span></a></li><li><a class="tocitem" href="#HUMAnN-for-multiple-samples"><span>HUMAnN for multiple samples</span></a></li><li><a class="tocitem" href="#Plotting"><span>Plotting</span></a></li><li><a class="tocitem" href="#Functions-and-types"><span>Functions and types</span></a></li></ul></li><li><a class="tocitem" href="../metaphlan/">Working with MetaPhlAn</a></li><li><a class="tocitem" href="../contributing/">Contributing</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Working with HUMAnN</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Working with HUMAnN</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/BiobakeryUtils.jl/blob/master/docs/src/humann.md" title="Edit on GitHub"><span class="docs-icon fab">ÔÇõ</span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="HUMAnN-Tutorial-with-BiobakeryUtils.jl"><a class="docs-heading-anchor" href="#HUMAnN-Tutorial-with-BiobakeryUtils.jl">HUMAnN Tutorial with BiobakeryUtils.jl</a><a id="HUMAnN-Tutorial-with-BiobakeryUtils.jl-1"></a><a class="docs-heading-anchor-permalink" href="#HUMAnN-Tutorial-with-BiobakeryUtils.jl" title="Permalink"></a></h1><ul><li>üóíÔ∏è This tutorial is meant to be run in parallel with / mirror the <a href="https://github.com/biobakery/biobakery/wiki/humann3">official HUMAnN v3 tutorial</a></li><li>‚ùìÔ∏è If you have questions about MetaPhlAn itself, please direct them to the <a href="https://forum.biobakery.org/c/Microbial-community-profiling/HUMAnN">bioBakery help forum</a></li><li>ü§î If you have questions about using the MetaPhlAn tools in julia, <a href="https://github.com/BioJulia/BiobakeryUtils.jl/issues/new/choose">please open an issue</a>, or start a discussion over on <a href="https://github.com/BioJulia/Microbiome.jl/discussions/new"><code>Microbiome.jl</code></a>!</li><li>üìî For a function / type reference, <a href="#Functions-and-Types">jump to the bottom</a></li></ul><h2 id="Installation-and-setup"><a class="docs-heading-anchor" href="#Installation-and-setup">Installation and setup</a><a id="Installation-and-setup-1"></a><a class="docs-heading-anchor-permalink" href="#Installation-and-setup" title="Permalink"></a></h2><p>If you haven&#39;t already, check out the <a href="gettingstarted">&quot;Getting Started&quot;</a> page to install julia, create an environment and install BiobakeryUtils.jl, and hook up or install the MetaPhlAn v3 command line tools.</p><p>This tutorial assumes:</p><ol><li>You are running julia v1.6 or greater</li><li>You have activated a julia Project that has <code>BiobakeryUtils.jl</code> installed</li><li>The <code>humann</code> python package is installed, and accessible from your <code>PATH</code>.</li></ol><p>If any of those things aren&#39;t true, or you don&#39;t know if they&#39;re true, go back to <a href="gettingstarted">&quot;Getting Started&quot;</a> to see if you skipped a step. If you&#39;re still confused, please ask (see 3rd bullet point at the top)!</p><h3 id="HUMAnN-Databases"><a class="docs-heading-anchor" href="#HUMAnN-Databases">HUMAnN Databases</a><a id="HUMAnN-Databases-1"></a><a class="docs-heading-anchor-permalink" href="#HUMAnN-Databases" title="Permalink"></a></h3><p>HUMAnN requires a number of specialized databases to work correctly. When you first install it, it comes with some demo databases that are much smaller, but can be used to complete this tutorial. However, for actually running real data, you&#39;ll want to take the time to download them - they&#39;re BIG! <a href="https://github.com/biobakery/humann#5-download-the-databases">See here</a> for more information.</p><p>For now, the easiest way to do this for now is via the shell, which you can access from the julia REPL by typing <code>;</code>:</p><pre><code class="language-julia-repl hljs">shell&gt; humann_databases --help
usage: humann_databases [-h] [--available]
                        [--download &lt;database&gt; &lt;build&gt; &lt;install_location&gt;]
                        [--update-config {yes,no}]
                        [--database-location DATABASE_LOCATION]

HUMAnN Databases

optional arguments:
  -h, --help            show this help message and exit
  --available           print the available databases
  --download &lt;database&gt; &lt;build&gt; &lt;install_location&gt;
                        download the selected database to the install location
  --update-config {yes,no}
                        update the config file to set the new database as the default [DEFAULT: yes]
  --database-location DATABASE_LOCATION
                        location (local or remote) to pull the database

shell&gt; humann_databases --available
HUMANnN2 Databases ( database : build = location )
chocophlan : full = http://huttenhower.sph.harvard.edu/humann_data/chocophlan/full_chocophlan.v296_201901b.tar.gz
chocophlan : DEMO = http://huttenhower.sph.harvard.edu/humann_data/chocophlan/DEMO_chocophlan.v296_201901b.tar.gz
uniref : uniref50_diamond = http://huttenhower.sph.harvard.edu/humann_data/uniprot/uniref_annotated/uniref50_annotated_v201901b_ful
l.tar.gz
uniref : uniref90_diamond = http://huttenhower.sph.harvard.edu/humann_data/uniprot/uniref_annotated/uniref90_annotated_v201901b_ful
l.tar.gz
uniref : uniref50_ec_filtered_diamond = http://huttenhower.sph.harvard.edu/humann_data/uniprot/uniref_ec_filtered/uniref50_ec_filte
red_201901b_subset.tar.gz
uniref : uniref90_ec_filtered_diamond = http://huttenhower.sph.harvard.edu/humann_data/uniprot/uniref_ec_filtered/uniref90_ec_filte
red_201901b_subset.tar.gz
uniref : DEMO_diamond = http://huttenhower.sph.harvard.edu/humann_data/uniprot/uniref_annotated/uniref90_DEMO_diamond_v201901b.tar.
gz
utility_mapping : full = http://huttenhower.sph.harvard.edu/humann_data/full_mapping_v201901b.tar.gz</code></pre><p>For example, if you&#39;d like to install these databases to <code>/BigDrive/humann/</code>, you could run</p><pre><code class="language-julia-repl hljs">shell&gt; humann_databases --download cholophlan full /BigDrive/humann/chocophlan
# ... lots of output

shell&gt; humann_databases --download uniref uniref90_diamond /BigDrive/humann/uniref
# ... lots of output

shell&gt; humann_databases --download utility_mapping full /BigDrive/humann/utility_mapping
# ... lots of output</code></pre><p>At some point, I&#39;ll write some functions to automate this, but for now, doing this will update a configuration file, so you shouldn&#39;t have to worry about it again.</p><h2 id="Running-HUMAnN"><a class="docs-heading-anchor" href="#Running-HUMAnN">Running HUMAnN</a><a id="Running-HUMAnN-1"></a><a class="docs-heading-anchor-permalink" href="#Running-HUMAnN" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#2-metagenome-functional-profiling">Official tutorial link</a></p><h2 id="Default-outputs"><a class="docs-heading-anchor" href="#Default-outputs">Default outputs</a><a id="Default-outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Default-outputs" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#23-humann-default-outputs">Official tutorial link</a></p><ul><li>load gene families file with <code>humann_profile</code></li><li>look at contents</li><li>filter on stratified, unstratified</li></ul><h2 id="Manipulating-tables"><a class="docs-heading-anchor" href="#Manipulating-tables">Manipulating tables</a><a id="Manipulating-tables-1"></a><a class="docs-heading-anchor-permalink" href="#Manipulating-tables" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#3-manipulating-humann-output-tables">Official tutorial link</a></p><ul><li>Do all of this in CommProfile</li></ul><h3 id="Normalize-RPK-to-relative-abundance"><a class="docs-heading-anchor" href="#Normalize-RPK-to-relative-abundance">Normalize RPK to relative abundance</a><a id="Normalize-RPK-to-relative-abundance-1"></a><a class="docs-heading-anchor-permalink" href="#Normalize-RPK-to-relative-abundance" title="Permalink"></a></h3><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#31-normalizing-rpks-to-relative-abundance">Official tutorial link</a></p><h3 id="Regrouping-genes-to-other-functional-categories"><a class="docs-heading-anchor" href="#Regrouping-genes-to-other-functional-categories">Regrouping genes to other functional categories</a><a id="Regrouping-genes-to-other-functional-categories-1"></a><a class="docs-heading-anchor-permalink" href="#Regrouping-genes-to-other-functional-categories" title="Permalink"></a></h3><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#32-regrouping-genes-to-other-functional-categories">Official tutorial link</a></p><ul><li>use <code>humann_regroup</code></li></ul><h3 id="Attaching-names-to-features"><a class="docs-heading-anchor" href="#Attaching-names-to-features">Attaching names to features</a><a id="Attaching-names-to-features-1"></a><a class="docs-heading-anchor-permalink" href="#Attaching-names-to-features" title="Permalink"></a></h3><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#33-attaching-names-to-features">Official tutorial link</a></p><ul><li>use <code>humann_rename</code></li></ul><h2 id="HUMAnN-for-multiple-samples"><a class="docs-heading-anchor" href="#HUMAnN-for-multiple-samples">HUMAnN for multiple samples</a><a id="HUMAnN-for-multiple-samples-1"></a><a class="docs-heading-anchor-permalink" href="#HUMAnN-for-multiple-samples" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#42-humann-for-multiple-samples">Official tutorial link</a></p><h2 id="Plotting"><a class="docs-heading-anchor" href="#Plotting">Plotting</a><a id="Plotting-1"></a><a class="docs-heading-anchor-permalink" href="#Plotting" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#5-plotting-stratified-functions">Official tutorial link</a></p><ul><li>Kevin should do this</li></ul><h2 id="Functions-and-types"><a class="docs-heading-anchor" href="#Functions-and-types">Functions and types</a><a id="Functions-and-types-1"></a><a class="docs-heading-anchor-permalink" href="#Functions-and-types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.humann" href="#BiobakeryUtils.humann"><code>BiobakeryUtils.humann</code></a> ‚Äî <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">humann(inputfile, output[, flags]; kwargs...)</code></pre><p>Run <code>humann</code> command line tool on <code>inputfile</code>, putting outputs in <code>output</code>. Requires <code>humann</code> to be installed and accessible in the <code>PATH</code> (see <a href="../gettingstarted/#Getting-Started">Getting Started</a>).</p><p><code>humann</code> flag options (those that don&#39;t have a parameter) can be passed in an array, and other options can be passed with keyword arguments. For example, if on the command line you would run:</p><pre><code class="language-sh hljs">$ humann -i $INPUTFILE -o $OUTPUT --bypass-tranlated-search --input-format fastq.gz --output-format biom</code></pre><p>using this function, you would write:</p><pre><code class="language-julia hljs">humann(INTPUTFILE, OUTPUT, [&quot;bypass_translated_search&quot;]; input_formal=&quot;fastq.gz&quot;, output_format=&quot;biom&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BiobakeryUtils.jl/blob/8f2b5a670ecc5e08dccad948473d51c8644b2872/src/humann.jl#L1-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.humann_profile-Tuple{AbstractString}" href="#BiobakeryUtils.humann_profile-Tuple{AbstractString}"><code>BiobakeryUtils.humann_profile</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><p>Stratified import currently non-functional</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BiobakeryUtils.jl/blob/8f2b5a670ecc5e08dccad948473d51c8644b2872/src/humann.jl#L38-L40">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.humann_profiles-Tuple{AbstractString}" href="#BiobakeryUtils.humann_profiles-Tuple{AbstractString}"><code>BiobakeryUtils.humann_profiles</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><p>Stratified import currently non-functional</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BiobakeryUtils.jl/blob/8f2b5a670ecc5e08dccad948473d51c8644b2872/src/humann.jl#L73-L75">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../gettingstarted/">¬´ Getting Started</a><a class="docs-footer-nextpage" href="../metaphlan/">Working with MetaPhlAn ¬ª</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.7 on <span class="colophon-date" title="Tuesday 5 October 2021 21:13">Tuesday 5 October 2021</span>. Using Julia version 1.6.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
