<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Working with HUMAnN ¬∑ BiobakeryUtils.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="http://docs.ecojulia.org/BiobakeryUtils.jl/stable/humann/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">BiobakeryUtils.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">BiobakeryUtils</a></li><li><a class="tocitem" href="../gettingstarted/">Getting Started</a></li><li><a class="tocitem" href="../kneaddata/">Working with KneadData</a></li><li><a class="tocitem" href="../metaphlan/">Working with MetaPhlAn</a></li><li class="is-active"><a class="tocitem" href>Working with HUMAnN</a><ul class="internal"><li><a class="tocitem" href="#Installation-and-setup"><span>Installation and setup</span></a></li><li><a class="tocitem" href="#Running-HUMAnN"><span>Running HUMAnN</span></a></li><li><a class="tocitem" href="#Default-outputs"><span>Default outputs</span></a></li><li><a class="tocitem" href="#Manipulating-tables"><span>Manipulating tables</span></a></li><li><a class="tocitem" href="#HUMAnN-for-multiple-samples"><span>HUMAnN for multiple samples</span></a></li><li><a class="tocitem" href="#Plotting"><span>Plotting</span></a></li><li><a class="tocitem" href="#Functions-and-types"><span>Functions and types</span></a></li></ul></li><li><a class="tocitem" href="../microbiome/">Microbiome.jl Docstrings</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Working with HUMAnN</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Working with HUMAnN</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/main/docs/src/humann.md" title="Edit on GitHub"><span class="docs-icon fab">ÔÇõ</span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="humann-tutorial"><a class="docs-heading-anchor" href="#humann-tutorial">HUMAnN Tutorial with BiobakeryUtils.jl</a><a id="humann-tutorial-1"></a><a class="docs-heading-anchor-permalink" href="#humann-tutorial" title="Permalink"></a></h1><ul><li>üóíÔ∏è This tutorial is meant to be run in parallel with / mirror the <a href="https://github.com/biobakery/biobakery/wiki/humann3">official HUMAnN v3 tutorial</a></li><li>‚ùìÔ∏è If you have questions about HUMAnN itself, please direct them to the <a href="https://forum.biobakery.org/c/Microbial-community-profiling/HUMAnN">bioBakery help forum</a></li><li>ü§î If you have questions about using the MetaPhlAn tools in julia, <a href="https://github.com/EcoJulia/BiobakeryUtils.jl/issues/new/choose">please open an issue</a>, or start a discussion over on <a href="https://github.com/EcoJulia/Microbiome.jl/discussions/new"><code>Microbiome.jl</code></a>!</li><li>üìî For a function / type reference, <a href="#Functions-and-Types">jump to the bottom</a></li></ul><h2 id="Installation-and-setup"><a class="docs-heading-anchor" href="#Installation-and-setup">Installation and setup</a><a id="Installation-and-setup-1"></a><a class="docs-heading-anchor-permalink" href="#Installation-and-setup" title="Permalink"></a></h2><p>If you haven&#39;t already, check out the <a href="../gettingstarted/#getting-started">&quot;Getting Started&quot;</a> page to install julia, create an environment, and install BiobakeryUtils.jl, and hook up or install the HUMAnN v3 command line tools.</p><p>This tutorial assumes:</p><ol><li>You are running julia v1.6 or greater</li><li>You have activated a julia Project that has <code>BiobakeryUtils.jl</code> installed</li><li>The <code>humann</code> python package is installed, and accessible from your <code>PATH</code>.</li></ol><p>If any of those things aren&#39;t true, or you don&#39;t know if they&#39;re true, go back to <a href="../gettingstarted/#getting-started">&quot;Getting Started&quot;</a> to see if you skipped a step. If you&#39;re still confused, please ask (see 3rd bullet point at the top)!</p><h3 id="HUMAnN-Databases"><a class="docs-heading-anchor" href="#HUMAnN-Databases">HUMAnN Databases</a><a id="HUMAnN-Databases-1"></a><a class="docs-heading-anchor-permalink" href="#HUMAnN-Databases" title="Permalink"></a></h3><p>HUMAnN requires a number of specialized databases to work correctly. When you first install it, it comes with some demo databases that are much smaller, but can be used to complete this tutorial. However, for actually running real data, you&#39;ll want to take the time to download them - they&#39;re BIG! <a href="https://github.com/biobakery/humann#5-download-the-databases">See here</a> for more information.</p><p>For now, the easiest way to do this for now is via the shell, which you can access from the julia REPL by typing <code>;</code>:</p><pre><code class="language-julia-repl hljs">;humann_databases --help
usage: humann_databases [-h] [--available]
                        [--download &lt;database&gt; &lt;build&gt; &lt;install_location&gt;]
                        [--update-config {yes,no}]
                        [--database-location DATABASE_LOCATION]

HUMAnN Databases

optional arguments:
  -h, --help            show this help message and exit
  --available           print the available databases
  --download &lt;database&gt; &lt;build&gt; &lt;install_location&gt;
                        download the selected database to the install location
  --update-config {yes,no}
                        update the config file to set the new database as the default [DEFAULT: yes]
  --database-location DATABASE_LOCATION
                        location (local or remote) to pull the database

;humann_databases --available
HUMANnN2 Databases ( database : build = location )
chocophlan : full = http://huttenhower.sph.harvard.edu/humann_data/chocophlan/full_chocophlan.v296_201901b.tar.gz
chocophlan : DEMO = http://huttenhower.sph.harvard.edu/humann_data/chocophlan/DEMO_chocophlan.v296_201901b.tar.gz
uniref : uniref50_diamond = http://huttenhower.sph.harvard.edu/humann_data/uniprot/uniref_annotated/uniref50_annotated_v201901b_ful
l.tar.gz
uniref : uniref90_diamond = http://huttenhower.sph.harvard.edu/humann_data/uniprot/uniref_annotated/uniref90_annotated_v201901b_ful
l.tar.gz
uniref : uniref50_ec_filtered_diamond = http://huttenhower.sph.harvard.edu/humann_data/uniprot/uniref_ec_filtered/uniref50_ec_filte
red_201901b_subset.tar.gz
uniref : uniref90_ec_filtered_diamond = http://huttenhower.sph.harvard.edu/humann_data/uniprot/uniref_ec_filtered/uniref90_ec_filte
red_201901b_subset.tar.gz
uniref : DEMO_diamond = http://huttenhower.sph.harvard.edu/humann_data/uniprot/uniref_annotated/uniref90_DEMO_diamond_v201901b.tar.
gz
utility_mapping : full = http://huttenhower.sph.harvard.edu/humann_data/full_mapping_v201901b.tar.gz</code></pre><p>For example, if you&#39;d like to install these databases to <code>/BigDrive/humann/</code>, you could run</p><pre><code class="language-julia-repl hljs">shell&gt; humann_databases --download chocophlan full /BigDrive/humann/chocophlan
# ... lots of output

shell&gt; humann_databases --download uniref uniref90_diamond /BigDrive/humann/uniref
# ... lots of output

shell&gt; humann_databases --download utility_mapping full /BigDrive/humann/utility_mapping
# ... lots of output</code></pre><p>At some point, I&#39;ll write some functions to automate this, but for now, doing this will update a configuration file, so you shouldn&#39;t have to worry about it again.</p><h2 id="Running-HUMAnN"><a class="docs-heading-anchor" href="#Running-HUMAnN">Running HUMAnN</a><a id="Running-HUMAnN-1"></a><a class="docs-heading-anchor-permalink" href="#Running-HUMAnN" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#2-metagenome-functional-profiling">Official tutorial link</a></p><p>Some example files you can use to run this tutorial are available from the MetaPhlAn repo, and can be downloaded using the <code>Downloads</code> standard library in julia:</p><pre><code class="language-julia-repl hljs">julia&gt; using Downloads: download

julia&gt; base_url = &quot;https://github.com/biobakery/humann/raw/master/examples/&quot;;

julia&gt; files = [
           &quot;demo.fastq.gz&quot;,
           &quot;demo.sam&quot;,
           &quot;demo.m8&quot;
       ];

julia&gt; for file in files
           download(joinpath(base_url, file), file)
       end

julia&gt; readdir()
5-element Vector{String}:
 &quot;Manifest.toml&quot;
 &quot;Project.toml&quot;
 &quot;demo.fastq.gz&quot;
 &quot;demo.sam&quot;
 &quot;demo.m8&quot;</code></pre><p>For convenience, this package has the <a href="#BiobakeryUtils.humann-Tuple{Any, Any}"><code>humann()</code></a> function, which can be used in your julia scripts to build and call the <code>humann</code> command line tool.</p><p>For example, rather than call</p><pre><code class="language-sh hljs">$ humann --input demo.fastq.gz --output demo_fastq</code></pre><p>You can do</p><pre><code class="language-julia-repl hljs">julia&gt; humann(&quot;demo.fastq.gz&quot;, &quot;demo_fastq&quot;)
[ Info: Running command: humann -i demo.fastq.gz -o demo_fastq
Creating output directory: /home/kevin/my_project/demo_fastq
Output files will be written to: /home/kevin/my_project/demo_fastq
Decompressing gzipped file ...
# ... etc</code></pre><p>First, <code>humann</code> will run <code>metaphlan</code> to generate taxonomic profiles, then will use that taxonomic profile to run a custom gene search.</p><h2 id="Default-outputs"><a class="docs-heading-anchor" href="#Default-outputs">Default outputs</a><a id="Default-outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Default-outputs" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#23-humann-default-outputs">Official tutorial link</a></p><p>To load a profile generated by <code>humann</code>, use the <a href="#BiobakeryUtils.humann_profile-Tuple{AbstractString}"><code>humann_profile</code></a> function:</p><pre><code class="language-julia-repl hljs">julia&gt; gfs = humann_profile(&quot;demo_fastq/demo_genefamilies.tsv&quot;)
Microbiome.CommunityProfile{Float64, Microbiome.GeneFunction, Microbiome.MicrobiomeSample} with 589 things in 1 places

Sample names:
UNMAPPED, UniRef90_G1UL42, UniRef90_I9QXW8...UniRef90_A6LH06, UniRef90_D0TRR5

Feature names:
demo_genefamilies</code></pre><p>HUMAnN generates &quot;stratified&quot; gene function profiles -  in other words, each gene function is also split into the species that contributed it. By default, <code>human_profile</code> skips the stratified rows (they can get big!):</p><pre><code class="language-julia-repl hljs">julia&gt; first(features(gfs), 5)
5-element Vector{GeneFunction}:
 GeneFunction(&quot;UNMAPPED&quot;, missing)
 GeneFunction(&quot;UniRef90_G1UL42&quot;, missing)
 GeneFunction(&quot;UniRef90_I9QXW8&quot;, missing)
 GeneFunction(&quot;UniRef90_A0A174QBF2&quot;, missing)
 GeneFunction(&quot;UniRef90_A0A078RDY6&quot;, missing)</code></pre><p>The <code>missing</code> component of the <a href="../microbiome/#Microbiome.GeneFunction"><code>GeneFunction</code></a> means that these gene functions are not associated with a particular taxon.</p><p>If you want to hang onto the taxon information, use the keyword argument <code>stratified = true</code>:</p><pre><code class="language-julia-repl hljs">julia&gt; gfs_strat = humann_profile(&quot;demo_fastq/demo_genefamilies.tsv&quot;, stratified=true)
CommunityProfile{Float64, GeneFunction, MicrobiomeSample} with 1416 features in 1 samples

Feature names:
UNMAPPED, UniRef90_G1UL42, UniRef90_G1UL42...UniRef90_D0TRR5, UniRef90_D0TRR5

Sample names:
demo_genefamilies



julia&gt; first(features(gfs_strat), 5)
5-element Vector{GeneFunction}:
 GeneFunction(&quot;UNMAPPED&quot;, missing)
 GeneFunction(&quot;UniRef90_G1UL42&quot;, missing)
 GeneFunction(&quot;UniRef90_G1UL42&quot;, Taxon(&quot;Bacteroides_dorei&quot;, :species))
 GeneFunction(&quot;UniRef90_I9QXW8&quot;, missing)
 GeneFunction(&quot;UniRef90_I9QXW8&quot;, Taxon(&quot;Bacteroides_dorei&quot;, :species))</code></pre><p>Here, we can see that the uniref90 &quot;G1UL42&quot; was contributed by <em>Bacteroides dorei</em>.</p><p>The object returned by <code>humann_profile</code> is a <code>CommunityProfile</code> type from <a href="https://github.com/EcoJulia/Microbiome.jl"><code>Microbiome.jl</code></a>, and has a bunch of useful properties.</p><h3 id="Indexing"><a class="docs-heading-anchor" href="#Indexing">Indexing</a><a id="Indexing-1"></a><a class="docs-heading-anchor-permalink" href="#Indexing" title="Permalink"></a></h3><p>For <code>CommunityProfile</code>s, you can select features and samples with strings (or regular expressions) representing names of features (rows) or samples (columns):</p><pre><code class="language-julia-repl hljs">julia&gt; samplenames(gfs_strat)
1-element Vector{String}:
 &quot;demo_genefamilies&quot;

julia&gt; featurenames(gfs_strat)
1416-element Vector{String}:
 &quot;UNMAPPED&quot;
 &quot;UniRef90_G1UL42&quot;
 &quot;UniRef90_G1UL42&quot;
 &quot;UniRef90_I9QXW8&quot;
 ‚ãÆ
 &quot;UniRef90_A6LH06&quot;
 &quot;UniRef90_D0TRR5&quot;
 &quot;UniRef90_D0TRR5&quot;
 &quot;UniRef90_D0TRR5&quot;

julia&gt; slice = gfs_strat[&quot;UniRef90_A0A174NIB7&quot;, :]
CommunityProfile{Float64, GeneFunction, MicrobiomeSample} with 3
features in 1 samples

Feature names:
UniRef90_A0A174NIB7, UniRef90_A0A174NIB7, UniRef90_A0A174NIB7

Sample names:
demo_genefamilies



julia&gt; features(slice)
3-element Vector{GeneFunction}:
 GeneFunction(&quot;UniRef90_A0A174NIB7&quot;, missing)
 GeneFunction(&quot;UniRef90_A0A174NIB7&quot;, Taxon(&quot;Bacteroides_dorei&quot;, :species))
 GeneFunction(&quot;UniRef90_A0A174NIB7&quot;, Taxon(&quot;Bacteroides_vulgatus&quot;, :species))</code></pre><p>Using a string (or regular expression) to index will return all rows whose <code>name</code> matches, regardless of the taxon. If you just want a single value, you can use a <code>GeneFunction</code> directly:</p><pre><code class="language-julia-repl hljs">julia&gt; gfs_strat[GeneFunction(&quot;UniRef90_A0A174NIB7&quot;, &quot;s__Bacteroides_dorei&quot;), 1]
0.8271298594</code></pre><p>You can even pass an array of strings as the row index to get a slice with multiple gene functions:</p><pre><code class="language-julia-repl hljs">julia&gt; features(gfs_strat[[&quot;UniRef90_A0A174NIB7&quot;, &quot;UniRef90_A6L100&quot;], :])
5-element Vector{GeneFunction}:
 GeneFunction(&quot;UniRef90_A6L100&quot;, missing)
 GeneFunction(&quot;UniRef90_A6L100&quot;, Taxon(&quot;Bacteroides_vulgatus&quot;, :species))
 GeneFunction(&quot;UniRef90_D0TRR5&quot;, missing)
 GeneFunction(&quot;UniRef90_D0TRR5&quot;, Taxon(&quot;Bacteroides_vulgatus&quot;, :species))
 GeneFunction(&quot;UniRef90_D0TRR5&quot;, Taxon(&quot;Bacteroides_dorei&quot;, :species))</code></pre><p>For more information about indexing and accessing components of the data, see <a href="https://docs.ecojulia.org/Microbiome.jl/latest/profiles/#Indexing-and-selecting-1">the Microbiome.jl docs</a></p><h2 id="Manipulating-tables"><a class="docs-heading-anchor" href="#Manipulating-tables">Manipulating tables</a><a id="Manipulating-tables-1"></a><a class="docs-heading-anchor-permalink" href="#Manipulating-tables" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#3-manipulating-humann-output-tables">Official tutorial link</a></p><p>There are several ways to manipulate <code>CommunityProfile</code>s, both using julia and using utilities provided by <code>humann</code>.</p><p>One example of the former is using <code>filter</code>, which takes a boolean function as the first argument, and returns a new <code>CommunityProfile</code> containing only rows that returned <code>true</code>.</p><p>For example, given a stratified table like <code>gfs_strat</code>, if you want to get only rows that have a taxon associated with them, you can do:</p><pre><code class="language-julia-repl hljs">julia&gt; gfs_strat_only = filter(hastaxon, gfs_strat)
CommunityProfile{Float64, GeneFunction, MicrobiomeSample} with 827 features in 1 samples

Feature names:
UniRef90_G1UL42, UniRef90_I9QXW8, UniRef90_A0A174QBF2...UniRef90_D0TRR5, UniRef90_D0TRR5

Sample names:
demo_genefamilies</code></pre><p>Uh oh! We&#39;ve now lost the &quot;UNMAPPED&quot; row, which means that we won&#39;t have the reads that couldn&#39;t be mapped to a gene function represented. No matter, we can use julia&#39;s <a href="https://docs.julialang.org/en/v1/manual/functions/#man-anonymous-functions">anonymous function</a> (also sometimes called &quot;lambda function&quot;) syntax to roll our own function.</p><p>In the following example, <code>gf -&gt;</code> indicates a function that takes a single argument (in this case, our <code>GeneFunction</code>), then askes if it&#39;s <a href="../microbiome/#Microbiome.name-Tuple{AbstractSample}"><code>name</code></a> is &quot;UNMAPPED&quot; with <code>name(gf) == &quot;UNMAPPED&quot;</code>, OR (<code>||</code> is a short-circuiting OR operator) if it has a taxon:</p><pre><code class="language-julia-repl hljs">julia&gt; gfs_with_unmapped = filter(
                            gf-&gt; name(gf) == &quot;UNMAPPED&quot; || hastaxon(gf),
                            gfs_strat)
CommunityProfile{Float64, GeneFunction, MicrobiomeSample} with 828 features in 1 samples

Feature names:
UNMAPPED, UniRef90_G1UL42, UniRef90_I9QXW8...UniRef90_D0TRR5, UniRef90_D0TRR5

Sample names:
demo_genefamilies</code></pre><h3 id="Normalize-RPK-to-relative-abundance"><a class="docs-heading-anchor" href="#Normalize-RPK-to-relative-abundance">Normalize RPK to relative abundance</a><a id="Normalize-RPK-to-relative-abundance-1"></a><a class="docs-heading-anchor-permalink" href="#Normalize-RPK-to-relative-abundance" title="Permalink"></a></h3><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#31-normalizing-rpks-to-relative-abundance">Official tutorial link</a></p><p>Some <code>humann</code> utility scripts have convenience functions in <code>BiobakeryUtils.jl</code>. for example, if you want to renormalize your table into relative abundance, you could use <code>humann_renorm_table</code> from the command line, or call <a href="#BiobakeryUtils.humann_renorm-Tuple{CommunityProfile}"><code>humann_renorm</code></a>:</p><pre><code class="language-julia-repl hljs">julia&gt; gfs_renormed = humann_renorm(gfs_strat; units=&quot;relab&quot;)
Loading table from: /tmp/jl_9WL33H
  Treating /tmp/jl_9WL33H as stratified output, e.g. [&#39;UniRef90_G1UL42&#39;, &#39;Bacteroides_dorei&#39;]
CommunityProfile{Float64, GeneFunction, MicrobiomeSample} with 589 features in 1 samples

Feature names:
UNMAPPED, UniRef90_G1UL42, UniRef90_I9QXW8...UniRef90_A6LH06, UniRef90_D0TRR5

Sample names:
demo_genefamilies

julia&gt; abundances(gfs_strat[1:5, 1])
5√ó1 SparseArrays.SparseMatrixCSC{Float64, Int64} with 5 stored entries:
 17556.0
   333.333
   333.333
   333.333
   333.333

julia&gt; abundances(gfs_renormed[1:5, 1])
5√ó1 SparseArrays.SparseMatrixCSC{Float64, Int64} with 5 stored entries:
 0.665379
 0.0126335
 0.0126335
 0.00758008
 0.00631673</code></pre><h3 id="Regrouping-genes-to-other-functional-categories"><a class="docs-heading-anchor" href="#Regrouping-genes-to-other-functional-categories">Regrouping genes to other functional categories</a><a id="Regrouping-genes-to-other-functional-categories-1"></a><a class="docs-heading-anchor-permalink" href="#Regrouping-genes-to-other-functional-categories" title="Permalink"></a></h3><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#32-regrouping-genes-to-other-functional-categories">Official tutorial link</a></p><p>Similarly, if we want to regroup our uniref90s into another gene function category like ecs or KOs, we can use <a href="#BiobakeryUtils.humann_regroup-Tuple{CommunityProfile}"><code>humann_regroup</code></a></p><pre><code class="language-julia-repl hljs">julia&gt; gfs_rxn = humann_regroup(gfs_strat, inkind=&quot;uniref90&quot;, outkind=&quot;rxn&quot;)
Loading table from: /tmp/jl_SA9rCQ
  Treating /tmp/jl_SA9rCQ as stratified output, e.g. [&#39;UniRef90_G1UL42&#39;, &#39;Bacteroides_dorei&#39;]
Loading mapping file from: /home/kevin/.julia/conda/3/envs/biobakery/lib/python3.7/site-packages/humann/tools/../data/pathways/meta
cyc_reactions_level4ec_only.uniref.bz2
Original Feature Count: 589; Grouped 1+ times: 78 (13.2%); Grouped 2+ times: 20 (3.4%)
CommunityProfile{Float64, GeneFunction, MicrobiomeSample} with 174 features in 1 samples

Feature names:
UNMAPPED, UNGROUPED, 1.7.7.2-RXN...UDPNACETYLGLUCOSAMACYLTRANS-RXN, UROGENDECARBOX-RXN

Sample names:
demo_genefamilies

julia&gt; first(features(gfs_strat), 5)
5-element Vector{GeneFunction}:
 GeneFunction(&quot;UNMAPPED&quot;, missing)
 GeneFunction(&quot;UniRef90_G1UL42&quot;, missing)
 GeneFunction(&quot;UniRef90_G1UL42&quot;, Taxon(&quot;Bacteroides_dorei&quot;, :species))
 GeneFunction(&quot;UniRef90_I9QXW8&quot;, missing)
 GeneFunction(&quot;UniRef90_I9QXW8&quot;, Taxon(&quot;Bacteroides_dorei&quot;, :species))

julia&gt; first(features(gfs_rxn), 5)
5-element Vector{GeneFunction}:
 GeneFunction(&quot;UNMAPPED&quot;, missing)
 GeneFunction(&quot;UNGROUPED&quot;, missing)
 GeneFunction(&quot;1.7.7.2-RXN&quot;, missing)
 GeneFunction(&quot;1.8.1.4-RXN&quot;, missing)
 GeneFunction(&quot;2.4.1.135-RXN&quot;, missing)</code></pre><p>Note - to get other feature types, you may have to download the requisite databases using <code>humann_databases</code> at the command line. See <a href="../gettingstarted/#using-conda">Using Conda.jl</a></p><h3 id="Attaching-names-to-features"><a class="docs-heading-anchor" href="#Attaching-names-to-features">Attaching names to features</a><a id="Attaching-names-to-features-1"></a><a class="docs-heading-anchor-permalink" href="#Attaching-names-to-features" title="Permalink"></a></h3><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#33-attaching-names-to-features">Official tutorial link</a></p><p>You can attach names to features using <a href="#BiobakeryUtils.humann_rename-Tuple{CommunityProfile}"><code>humann_rename</code></a>.</p><h2 id="HUMAnN-for-multiple-samples"><a class="docs-heading-anchor" href="#HUMAnN-for-multiple-samples">HUMAnN for multiple samples</a><a id="HUMAnN-for-multiple-samples-1"></a><a class="docs-heading-anchor-permalink" href="#HUMAnN-for-multiple-samples" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#42-humann-for-multiple-samples">Official tutorial link</a></p><p>You can easily run multiple files in a loop in julia. First, download the files (if you already did this <a href="../metaphlan/#metaphlan-multi">in the MetaPhlAn tutorial</a>, no need to repeat it).</p><pre><code class="language-julia-repl hljs">julia&gt; base_url = &quot;https://github.com/biobakery/biobakery/raw/master/demos/biobakery_demos/data/metaphlan3/input/&quot;;

julia&gt; files = [
           &quot;SRS014476-Supragingival_plaque.fasta.gz&quot;,
           &quot;SRS014494-Posterior_fornix.fasta.gz&quot;,
           &quot;SRS014459-Stool.fasta.gz&quot;,
           &quot;SRS014464-Anterior_nares.fasta.gz&quot;,
           &quot;SRS014470-Tongue_dorsum.fasta.gz&quot;,
           &quot;SRS014472-Buccal_mucosa.fasta.gz&quot;
       ];

julia&gt; for file in files
           download(joinpath(base_url, file), file)
       end</code></pre><p>Then, just write a normal loop with <code>humann</code>:</p><pre><code class="language-julia-repl hljs">julia&gt; for file in files
           humann(file, &quot;hmp_subset&quot;)
       end
[ Info: Running command: humann -i SRS014476-Supragingival_plaque.fasta.gz -o hmp_subset
Creating output directory: /home/kevin/my_project/hmp_subset
Output files will be written to: /home/kevin/my_project/hmp_subset
Decompressing gzipped file ...
# ... etc</code></pre><p>On my decently powerful laptop, this took about 10 min.</p><p>To merge them using <code>humann_join_tables</code>, use the convenient julia function, <a href="#BiobakeryUtils.humann_join-Tuple{Any, Any}"><code>humann_join</code></a>:</p><pre><code class="language-julia-repl hljs">julia&gt; humann_join(&quot;hmp_subset&quot;, &quot;hmp_subset_genefamilies.tsv&quot;; file_name=&quot;genefamilies&quot;)
Gene table created: /home/kevin/my_project/hmp_subset_genefamilies.tsv
Process(`humann_join_tables -i hmp_subset -o hmp_subset_genefamilies.tsv --file_name genefamilies`, ProcessExited(0))</code></pre><p>This will write a new file that you can then load with <a href="#BiobakeryUtils.humann_profiles-Tuple{AbstractString}"><code>humann_profiles</code></a></p><pre><code class="language-julia-repl hljs">julia&gt; humann_profiles(&quot;hmp_subset_genefamilies.tsv&quot;; stratified=true)
</code></pre><p>Alternatively, you can load each profile into a <code>CommunityProfile</code>, then merge them using the <code>Microbiome.jl</code> function <code>commjoin</code>:</p><pre><code class="language-julia-repl hljs"># anonymous function passed to filter files that contain &quot;genefamilies&quot;
julia&gt; hmp_files = filter(f-&gt; contains(f, &quot;genefamilies&quot;),
                              readdir(&quot;hmp_subset&quot;; join=true))
6-element Vector{String}:
 &quot;hmp_subset/SRS014459-Stool_genefamilies.tsv&quot;
 &quot;hmp_subset/SRS014464-Anterior_nares_genefamilies.tsv&quot;
 &quot;hmp_subset/SRS014470-Tongue_dorsum_genefamilies.tsv&quot;
 &quot;hmp_subset/SRS014472-Buccal_mucosa_genefamilies.tsv&quot;
 &quot;hmp_subset/SRS014476-Supragingival_plaque_genefamilies.tsv&quot;
 &quot;hmp_subset/SRS014494-Posterior_fornix_genefamilies.tsv&quot;

julia&gt; commjoin(humann_profile.(hmp_files)...)
CommunityProfile{Float64, GeneFunction, MicrobiomeSample} with 1 features in 6 samples

Feature names:
UNMAPPED

Sample names:
SRS014459-Stool_genefamilies, SRS014464-Anterior_nares_genefamilies, SRS014470-Tongue_dorsum_genefamilies...SRS014476-Supragingival
_plaque_genefamilies, SRS014494-Posterior_fornix_genefamilies</code></pre><h2 id="Plotting"><a class="docs-heading-anchor" href="#Plotting">Plotting</a><a id="Plotting-1"></a><a class="docs-heading-anchor-permalink" href="#Plotting" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/humann3#5-plotting-stratified-functions">Official tutorial link</a></p><p><code>BiobakeryUtils.jl</code> does not come with plotting recipes (yet), but there are several excellent plotting packages that you can use. Alternatively, you can use the wrapped <a href="#BiobakeryUtils.humann_barplot-Tuple{CommunityProfile, Any}"><code>humann_barplot</code></a> script.</p><p>First, download the pcl file used in the HUMAnN tutorial.</p><pre><code class="language-julia-repl hljs">julia&gt; download(&quot;https://raw.githubusercontent.com/biobakery/biobakery/master/demos/biobakery_demos/data/humann2/input/hmp_pathabund.pcl&quot;, &quot;hmp_pathabund.pcl&quot;)</code></pre><h3 id="Using-humann_barplot"><a class="docs-heading-anchor" href="#Using-humann_barplot">Using humann_barplot</a><a id="Using-humann_barplot-1"></a><a class="docs-heading-anchor-permalink" href="#Using-humann_barplot" title="Permalink"></a></h3><p>It&#39;s probably a good idea to read the tutorial link above that describes the dataset. here are the equivalent julia commands to generate the plots described there.</p><pre><code class="language-julia-repl hljs">julia&gt; gfs = read_pcl(&quot;hmp_pathabund.pcl&quot;; last_metadata=&quot;STSite&quot;)



julia&gt; humann_barplot(gfs, &quot;plot1.png&quot;; focal_metadata=&quot;STSite&quot;, focal_feature=&quot;METSYN-PWY&quot;)
Process(`humann_barplot --i /tmp/jl_tKWIfW -o plot1.png --last-metadata STSite --focal-metadata STSite --focal-feature METSYN-PWY`,

julia&gt; humann_barplot(gfs, &quot;plot2.png&quot;; focal_metadata=&quot;STSite&quot;, focal_feature=&quot;METSYN-PWY&quot;, 
                      sort=&quot;sum&quot;)
Process(`humann_barplot --i /tmp/jl_vF6GHe -o plot2.png --last-metadata STSite --focal-metadata STSite --focal-feature METSYN-PWY -

julia&gt; humann_barplot(gfs, &quot;plot3.png&quot;; focal_metadata=&quot;STSite&quot;, focal_feature=&quot;METSYN-PWY&quot;,
                      sort=[&quot;sum&quot;, &quot;metadata&quot;],
                      scaling=&quot;logstack&quot;)
Process(`humann_barplot --i /tmp/jl_VVl3zD -o plot3.png --last-metadata STSite --focal-metadata STSite --focal-feature METSYN-PWY -
-sort sum metadata --scaling logstack`, ProcessExited(0))

julia&gt; humann_barplot(gfs, &quot;plot4.png&quot;; focal_metadata=&quot;STSite&quot;, focal_feature=&quot;COA-PWY&quot;,
                      sort=&quot;sum&quot;)
Process(`humann_barplot --i /tmp/jl_XePIBD -o plot4.png --last-metadata STSite --focal-metadata STSite --focal-feature COA-PWY --so
rt sum`, ProcessExited(0))

julia&gt; humann_barplot(gfs, &quot;plot5.png&quot;; focal_metadata=&quot;STSite&quot;, focal_feature=&quot;COA-PWY&quot;,
                      sort=&quot;braycurtis&quot;,
                      scaling=&quot;logstack&quot;,
                      as_genera=true,
                      remove_zeros=true)</code></pre><p>On the last call, notice that &quot;flag arguments&quot; (eg <code>--as-genera</code>) that don&#39;t take arguments on the command line must be set to <code>true</code> in the julia version.</p><h3 id="Using-julia-plotting"><a class="docs-heading-anchor" href="#Using-julia-plotting">Using julia plotting</a><a id="Using-julia-plotting-1"></a><a class="docs-heading-anchor-permalink" href="#Using-julia-plotting" title="Permalink"></a></h3><p>Use the <a href="#BiobakeryUtils.read_pcl-Tuple{Any}"><code>read_pcl</code></a> function to load the pcl file into julia, which will add all of the metadata encoded in the PCL to the resulting <code>CommunityProfile</code></p><pre><code class="language-julia-repl hljs">julia&gt; gfs = read_pcl(&quot;hmp_pathabund.pcl&quot;, last_metadata=&quot;STSite&quot;)
CommunityProfile{Float64, GeneFunction, MicrobiomeSample} with 5606 features in 378 samples

Feature names:
1CMET2-PWY: N10-formyl-tetrahydrofolate biosynthesis, 1CMET2-PWY: N10-formyl-tetrahydrofolate biosynthesis, 1CMET2-PWY: N10-formyl-
tetrahydrofolate biosynthesis...VALSYN-PWY: L-valine biosynthesis, VALSYN-PWY: L-valine biosynthesis

Sample names:
SRS011084, SRS011086, SRS011090...SRS058213, SRS058808



julia&gt; first(samples(gfs))
MicrobiomeSample(&quot;SRS011084&quot;, {:STSite = &quot;Stool&quot;})</code></pre><p>For plotting, I tend to use <a href="https://github.com/JuliaPlots/Makie.jl">Makie</a>, but there are <a href="https://juliahub.com/ui/Search?q=plotting&amp;type=packages">many other options</a>.</p><h2 id="Functions-and-types"><a class="docs-heading-anchor" href="#Functions-and-types">Functions and types</a><a id="Functions-and-types-1"></a><a class="docs-heading-anchor-permalink" href="#Functions-and-types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.humann-Tuple{Any, Any}" href="#BiobakeryUtils.humann-Tuple{Any, Any}"><code>BiobakeryUtils.humann</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">humann(inputfile, output[, flags]; kwargs...)</code></pre><p>Run <code>humann</code> command line tool on <code>inputfile</code>, putting outputs in <code>output</code>. Requires <code>humann</code> to be installed and accessible in the <code>PATH</code> (see <a href="@ref">Getting Started</a>).</p><p><code>humann</code> flag options (those that don&#39;t have a parameter) can be passed in an array, and other options can be passed with keyword arguments. For example, if on the command line you would run:</p><pre><code class="language-sh hljs">$ humann -i $INPUTFILE -o $OUTPUT --bypass-tranlated-search --input-format fastq.gz --output-format biom</code></pre><p>using this function, you would write:</p><pre><code class="language-julia hljs">humann(INTPUTFILE, OUTPUT, [&quot;bypass_translated_search&quot;]; input_formal=&quot;fastq.gz&quot;, output_format=&quot;biom&quot;)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/15c9512066b9677ab73ef0899a0bbfa862dbe7cb/src/humann.jl#L1-L22">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.humann_barplot-Tuple{CommunityProfile, Any}" href="#BiobakeryUtils.humann_barplot-Tuple{CommunityProfile, Any}"><code>BiobakeryUtils.humann_barplot</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">humann_barplot(comm::CommunityProfile, outpath; kwargs...)</code></pre><p>Wrapper for <code>humann_barplot</code> script, to generate plots from functional data. pass keyword arguments for script options. Flag arguments should be set to <code>true</code>. eg</p><pre><code class="language-julia-repl hljs">julia&gt; humann_barplot(comm, &quot;plot.png&quot;; focal_metadata=&quot;STSite&quot;, focal_feature=&quot;COA-PWY&quot;,
                      sort=&quot;braycurtis&quot;,
                      scaling=&quot;logstack&quot;,
                      as_genera=true,
                      remove_zeros=true)</code></pre><p>Requires installation of <a href="https://github.com/biobakery/humann"><code>humann</code></a> available in <code>ENV[&quot;PATH&quot;]</code>. See &quot;<a href="../gettingstarted/#using-conda">Using Conda</a>&quot; for more information.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/15c9512066b9677ab73ef0899a0bbfa862dbe7cb/src/humann.jl#L247-L265">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.humann_join-Tuple{Any, Any}" href="#BiobakeryUtils.humann_join-Tuple{Any, Any}"><code>BiobakeryUtils.humann_join</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">humann_renorm(comm::AbstractDataFrame; units::String=&quot;cpm&quot;)</code></pre><p>Wrapper for <code>humann_renorm_table</code> script, to renormalize from RPKM (reads per kilobase per million) to &quot;cpm&quot; (counts per million) or &quot;relab&quot; (relative abundance).</p><p>Requires installation of <a href="https://github.com/biobakery/humann"><code>humann</code></a> available in <code>ENV[&quot;PATH&quot;]</code>. See &quot;<a href="../gettingstarted/#using-conda">Using Conda</a>&quot; for more information.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/15c9512066b9677ab73ef0899a0bbfa862dbe7cb/src/humann.jl#L173-L182">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.humann_profile-Tuple{AbstractString}" href="#BiobakeryUtils.humann_profile-Tuple{AbstractString}"><code>BiobakeryUtils.humann_profile</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">humann_profile(path::AbstractString; sample=basename(first(splitext(path))), stratified=false)</code></pre><p>Load a single functional profile generated by <code>HUMAnN</code>. By default, skips rows that have species-stratified content, use <code>stratified=true</code> to keep them.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/15c9512066b9677ab73ef0899a0bbfa862dbe7cb/src/humann.jl#L56-L62">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.humann_profiles-Tuple{AbstractString}" href="#BiobakeryUtils.humann_profiles-Tuple{AbstractString}"><code>BiobakeryUtils.humann_profiles</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div></div><a class="docs-sourcelink" target="_blank" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/15c9512066b9677ab73ef0899a0bbfa862dbe7cb/src/humann.jl#L78-L80">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.humann_regroup-Tuple{CommunityProfile}" href="#BiobakeryUtils.humann_regroup-Tuple{CommunityProfile}"><code>BiobakeryUtils.humann_regroup</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">function humann_regroup(comm::CommunityProfile; inkind=&quot;uniref90&quot;, outkind::String=&quot;ec&quot;)</code></pre><p>Wrapper for <code>humann_regroup_table</code> script to convert table from one kind of functional mapping to another.</p><p>Requires installation of <a href="https://github.com/biobakery/humann"><code>humann</code></a> available in <code>ENV[&quot;PATH&quot;]</code>. See &quot;<a href="../gettingstarted/#using-conda">Using Conda</a>&quot; for more information.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/15c9512066b9677ab73ef0899a0bbfa862dbe7cb/src/humann.jl#L103-L111">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.humann_rename-Tuple{CommunityProfile}" href="#BiobakeryUtils.humann_rename-Tuple{CommunityProfile}"><code>BiobakeryUtils.humann_rename</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">humann_rename(comm::AbstractDataFrame; kind::String=&quot;ec&quot;)</code></pre><p>Wrapper for <code>humann_rename_table</code> script, returning a CommunityProfile with re-named features.</p><p>Requires installation of <a href="https://github.com/biobakery/humann"><code>humann</code></a> available in <code>ENV[&quot;PATH&quot;]</code>. See &quot;<a href="../gettingstarted/#using-conda">Using Conda</a>&quot; for more information.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/15c9512066b9677ab73ef0899a0bbfa862dbe7cb/src/humann.jl#L126-L134">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.humann_renorm-Tuple{CommunityProfile}" href="#BiobakeryUtils.humann_renorm-Tuple{CommunityProfile}"><code>BiobakeryUtils.humann_renorm</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">humann_renorm(comm::AbstractDataFrame; units::String=&quot;cpm&quot;)</code></pre><p>Wrapper for <code>humann_renorm_table</code> script, to renormalize from RPKM (reads per kilobase per million) to &quot;cpm&quot; (counts per million) or &quot;relab&quot; (relative abundance).</p><p>Requires installation of <a href="https://github.com/biobakery/humann"><code>humann</code></a> available in <code>ENV[&quot;PATH&quot;]</code>. See &quot;<a href="../gettingstarted/#using-conda">Using Conda</a>&quot; for more information.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/15c9512066b9677ab73ef0899a0bbfa862dbe7cb/src/humann.jl#L149-L158">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.read_pcl-Tuple{Any}" href="#BiobakeryUtils.read_pcl-Tuple{Any}"><code>BiobakeryUtils.read_pcl</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">read_pcl(infile; last_metadata=2)</code></pre><p>Reads a <a href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#PCL:_Stanford_cDNA_file_format_.28.2A.pcl.29">PCL file</a> and generates a <code>CommunityProfile</code> with metadata attached to the samples.</p><p><code>last_metadata</code> may be a row number or a string representing the final metadatum.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/15c9512066b9677ab73ef0899a0bbfa862dbe7cb/src/humann.jl#L193-L200">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.write_pcl-Tuple{Any, CommunityProfile}" href="#BiobakeryUtils.write_pcl-Tuple{Any, CommunityProfile}"><code>BiobakeryUtils.write_pcl</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">write_pcl(infile; usemetadata=:all)</code></pre><p>Writes a <a href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#PCL:_Stanford_cDNA_file_format_.28.2A.pcl.29">PCL file</a> from a <code>CommunityProfile</code> with metadata attached to the samples.</p><p><code>usemetadata</code> may be <code>:all</code>  or a vector of symbols.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/EcoJulia/BiobakeryUtils.jl/blob/15c9512066b9677ab73ef0899a0bbfa862dbe7cb/src/humann.jl#L224-L232">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../metaphlan/">¬´ Working with MetaPhlAn</a><a class="docs-footer-nextpage" href="../microbiome/">Microbiome.jl Docstrings ¬ª</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Wednesday 1 February 2023 22:05">Wednesday 1 February 2023</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
