<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Working with MetaPhlAn ¬∑ BiobakeryUtils.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">BiobakeryUtils.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">BiobakeryUtils</a></li><li><a class="tocitem" href="../gettingstarted/">Getting Started</a></li><li><a class="tocitem" href="../general/">General Utilities</a></li><li><a class="tocitem" href="../humann/">Working with HUMAnN</a></li><li class="is-active"><a class="tocitem" href>Working with MetaPhlAn</a><ul class="internal"><li><a class="tocitem" href="#Installation-and-setup"><span>Installation and setup</span></a></li><li><a class="tocitem" href="#Input-files"><span>Input files</span></a></li><li><a class="tocitem" href="#Run-a-single-sample"><span>Run a single sample</span></a></li><li><a class="tocitem" href="#Output-files"><span>Output files</span></a></li><li><a class="tocitem" href="#Run-on-multiple-cores"><span>Run on multiple cores</span></a></li><li><a class="tocitem" href="#Run-multiple-samples"><span>Run multiple samples</span></a></li><li><a class="tocitem" href="#Merge-outputs"><span>Merge outputs</span></a></li><li><a class="tocitem" href="#Visualize-results"><span>Visualize results</span></a></li><li><a class="tocitem" href="#Functions-and-Types"><span>Functions and Types</span></a></li></ul></li><li><a class="tocitem" href="../contributing/">Contributing</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Working with MetaPhlAn</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Working with MetaPhlAn</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/BiobakeryUtils.jl/blob/master/docs/src/metaphlan.md" title="Edit on GitHub"><span class="docs-icon fab">ÔÇõ</span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="MetaPhlAn-Tutorial-with-BiobakeryUtils.jl"><a class="docs-heading-anchor" href="#MetaPhlAn-Tutorial-with-BiobakeryUtils.jl">MetaPhlAn Tutorial with BiobakeryUtils.jl</a><a id="MetaPhlAn-Tutorial-with-BiobakeryUtils.jl-1"></a><a class="docs-heading-anchor-permalink" href="#MetaPhlAn-Tutorial-with-BiobakeryUtils.jl" title="Permalink"></a></h1><ul><li>üóíÔ∏è This tutorial is meant to be run in parallel with / mirror the <a href="https://github.com/biobakery/biobakery/wiki/metaphlan3">official MetaPhlAn v3 tutorial</a></li><li>‚ùìÔ∏è If you have questions about MetaPhlAn itself, please direct them to the <a href="https://forum.biobakery.org/c/Microbial-community-profiling/MetaPhlAn">bioBakery help forum</a></li><li>ü§î If you have questions about using the MetaPhlAn tools in julia, <a href="https://github.com/BioJulia/BiobakeryUtils.jl/issues/new/choose">please open an issue</a>), or start a discussion over on <a href="https://github.com/BioJulia/Microbiome.jl/discussions/new"><code>Microbiome.jl</code></a>)!</li><li>üìî For a function / type reference, <a href="#Functions-and-Types">jump to the bottom</a></li></ul><h2 id="Installation-and-setup"><a class="docs-heading-anchor" href="#Installation-and-setup">Installation and setup</a><a id="Installation-and-setup-1"></a><a class="docs-heading-anchor-permalink" href="#Installation-and-setup" title="Permalink"></a></h2><p>If you haven&#39;t already, check out the <a href="gettingstarted">&quot;Getting Started&quot;</a> page to install julia, create an environment and install BiobakeryUtils.jl, and hook up or install the MetaPhlAn v3 command line tools.</p><p>This tutorial assumes:</p><ol><li>You are running julia v1.6 or greater</li><li>You have activated a julia Project that has <code>BiobakeryUtils.jl</code> installed</li><li>The <code>metaphlan</code> python package is installed, and accessible from your <code>PATH</code>.</li></ol><p>If any of those things aren&#39;t true, or you don&#39;t know if they&#39;re true, go back to <a href="gettingstarted">&quot;Getting Started&quot;</a> to see if you skipped a step. If you&#39;re still confused, please ask (see 3rd bullet point at the top)!</p><h3 id="Bowtie2-database"><a class="docs-heading-anchor" href="#Bowtie2-database">Bowtie2 database</a><a id="Bowtie2-database-1"></a><a class="docs-heading-anchor-permalink" href="#Bowtie2-database" title="Permalink"></a></h3><p>The first time you run <code>metaphlan</code>, it needs to download and unpack the marker database. If you don&#39;t care where this goes, don&#39;t worry about it - by default it will go into a subdirectory of your <code>conda</code> environment.</p><p>If your home folder has limited space, or you want to install it to a particular location (eg a faster drive), you can either pass the kewword argument <code>bowtie2db=&quot;/path/to/location&quot;</code> to all <code>metaphlan</code> commands, or set the environment variable <code>METAPHLAN_BOWTIE2_DB</code>.</p><h2 id="Input-files"><a class="docs-heading-anchor" href="#Input-files">Input files</a><a id="Input-files-1"></a><a class="docs-heading-anchor-permalink" href="#Input-files" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/metaphlan3#input-files">Official tutorial link</a></p><p>Some example files you can use to run this tutorial are available from the MetaPhlAn repo, and can be downloaded using the <code>Downloads</code> standard library in julia:</p><pre><code class="language-julia-repl hljs">julia&gt; using Downloads: download

julia&gt; base_url = &quot;https://github.com/biobakery/biobakery/raw/master/demos/biobakery_demos/data/metaphlan3/input/&quot;;

julia&gt; files = [
    &quot;SRS014476-Supragingival_plaque.fasta.gz&quot;,
    &quot;SRS014494-Posterior_fornix.fasta.gz&quot;,
    &quot;SRS014459-Stool.fasta.gz&quot;,
    &quot;SRS014464-Anterior_nares.fasta.gz&quot;,
    &quot;SRS014470-Tongue_dorsum.fasta.gz&quot;,
    &quot;SRS014472-Buccal_mucosa.fasta.gz&quot;
];

julia&gt; for file in files
           download(joinpath(base_url, file), file)
       end

julia&gt; readdir()
8-element Vector{String}:
 &quot;Manifest.toml&quot;
 &quot;Project.toml&quot;
 &quot;SRS014459-Stool.fasta.gz&quot;
 &quot;SRS014464-Anterior_nares.fasta.gz&quot;
 &quot;SRS014470-Tongue_dorsum.fasta.gz&quot;
 &quot;SRS014472-Buccal_mucosa.fasta.gz&quot;
 &quot;SRS014476-Supragingival_plaque.fasta.gz&quot;
 &quot;SRS014494-Posterior_fornix.fasta.gz&quot;</code></pre><h2 id="Run-a-single-sample"><a class="docs-heading-anchor" href="#Run-a-single-sample">Run a single sample</a><a id="Run-a-single-sample-1"></a><a class="docs-heading-anchor-permalink" href="#Run-a-single-sample" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/metaphlan3#run-a-single-sample">Official tutorial link</a></p><p>For convenience, this package has the <a href="#BiobakeryUtils.metaphlan-Tuple{Any, Any}"><code>metaphlan()</code></a> function, which can be used in your julia scripts to build and call the <code>metaphlan</code> command line tool.</p><p>For example, rather than call</p><pre><code class="language-sh hljs">$ metaphlan SRS014476-Supragingival_plaque.fasta.gz --input_type fasta &gt; SRS014476-Supragingival_plaque_profile.txt</code></pre><p>from the julia REPL, you can call</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; metaphlan(&quot;SRS014476-Supragingival_plaque.fasta.gz&quot;, &quot;SRS014476-Supragingival_plaque_profile.tsv&quot;; input_type=&quot;fasta&quot;)</code><code class="nohighlight hljs ansi" style="display:block;">ERROR: UndefVarError: metaphlan not defined</code></pre><h2 id="Output-files"><a class="docs-heading-anchor" href="#Output-files">Output files</a><a id="Output-files-1"></a><a class="docs-heading-anchor-permalink" href="#Output-files" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/metaphlan3#output-files">Official tutorial link</a></p><ul><li>load output with <code>metaphlan_profile</code></li><li>investigate with various functions (try to show similar things as tutorial)</li></ul><h2 id="Run-on-multiple-cores"><a class="docs-heading-anchor" href="#Run-on-multiple-cores">Run on multiple cores</a><a id="Run-on-multiple-cores-1"></a><a class="docs-heading-anchor-permalink" href="#Run-on-multiple-cores" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/metaphlan3#run-on-multiple-cores">Official tutorial link</a></p><h2 id="Run-multiple-samples"><a class="docs-heading-anchor" href="#Run-multiple-samples">Run multiple samples</a><a id="Run-multiple-samples-1"></a><a class="docs-heading-anchor-permalink" href="#Run-multiple-samples" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/metaphlan3#run-multiple-samples">Official tutorial link</a></p><h2 id="Merge-outputs"><a class="docs-heading-anchor" href="#Merge-outputs">Merge outputs</a><a id="Merge-outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Merge-outputs" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/metaphlan3#merge-outputs">Official tutorial link</a></p><ul><li>use <code>metaphlan_profile</code> in loop and then <code>commjoin</code></li><li>use <code>metaphlan_profiles</code></li></ul><h2 id="Visualize-results"><a class="docs-heading-anchor" href="#Visualize-results">Visualize results</a><a id="Visualize-results-1"></a><a class="docs-heading-anchor-permalink" href="#Visualize-results" title="Permalink"></a></h2><p><a href="https://github.com/biobakery/biobakery/wiki/metaphlan3#visualize-results">Official tutorial link</a></p><ul><li>Kevin should probably handle this part</li></ul><h2 id="Functions-and-Types"><a class="docs-heading-anchor" href="#Functions-and-Types">Functions and Types</a><a id="Functions-and-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Functions-and-Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.metaphlan-Tuple{Any, Any}" href="#BiobakeryUtils.metaphlan-Tuple{Any, Any}"><code>BiobakeryUtils.metaphlan</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">metaphlan(inputfile, outputfile; kwargs...)</code></pre><p>Run <code>metaphlan</code> command line tool on <code>inputfile</code>, creating <code>output</code>. Requires <code>metaphlan</code> to be installed and accessible in the <code>PATH</code> (see <a href="../gettingstarted/#Getting-Started">Getting Started</a>).</p><p><code>metaphlan</code> options can be passed via keyword arguments. For example, if on the command line you would run:</p><pre><code class="language-sh hljs">$ metaphlan some.fastq.gz output/some_profile.tsv --input_type fastq --nprocs 8</code></pre><p>using this function, you would write:</p><pre><code class="language-julia hljs">metaphlan(&quot;some.fastq.gz&quot;, &quot;output/some_profile.tsv&quot;; input_type=&quot;fastq&quot;, nprocs=8)</code></pre><p>Note: the <code>input_type</code> keyword is required.</p><p>Set the environmental variable &quot;METAPHLAN<em>BOWTIE2</em>DB&quot; to specify the location where the markergene database is/will be installed, or pass <code>bowtie2db = &quot;some/path&quot;</code> as a keyword argument.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BiobakeryUtils.jl/blob/971791c639d54143c90ad2743182bfaf5917376a/src/metaphlan.jl#L5-L31">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.metaphlan_profile" href="#BiobakeryUtils.metaphlan_profile"><code>BiobakeryUtils.metaphlan_profile</code></a> ‚Äî <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">metaphlan_profile(path::AbstractString, level::Union{Int, Symbol}=:all; sample::AbstractString=basename(first(splitext(path))))</code></pre><p>Compiles a MetaPhlAn file into a CommunityProfile. Can select data according to taxonomic level. If level not given, all data is compiled. <code>Sample name</code> of the CommunityProfile can be specified by passing a <code>sample</code> argument. If name not given, the name of the file becomes the <code>Sample name</code>.</p><p>Levels may be given either as numbers or symbols:</p><ul><li><code>1</code> = <code>:kingdom</code></li><li><code>2</code> = <code>:phylum</code></li><li><code>3</code> = <code>:class</code></li><li><code>4</code> = <code>:order</code></li><li><code>5</code> = <code>:family</code></li><li><code>6</code> = <code>:genus</code></li><li><code>7</code> = <code>:species</code></li><li><code>8</code> = <code>:subspecies</code></li></ul><pre><code class="language-julia-repl hljs">
Examples
‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°
julia&gt; metaphlan_profile(&quot;test/files/metaphlan_single2.tsv&quot;)
CommunityProfile{Float64, Taxon, MicrobiomeSample} with 96 features in 1 samples

Feature names:
Bacteria, Archaea, Firmicutes...Ruminococcus_bromii, Bacteroides_vulgatus

Sample names:
metaphlan_single2



julia&gt; metaphlan_profile(&quot;test/files/metaphlan_single2.tsv&quot;, 4)
CommunityProfile{Float64, Taxon, MicrobiomeSample} with 11 features in 1 samples

Feature names:
Clostridiales, Bacteroidales, Coriobacteriales...Firmicutes_unclassified, Pasteurellales

Sample names:
metaphlan_single2



julia&gt; metaphlan_profile(&quot;test/files/metaphlan_single2.tsv&quot;, :genus)
CommunityProfile{Float64, Taxon, MicrobiomeSample} with 40 features in 1 samples

Feature names:
Prevotella, Roseburia, Faecalibacterium...Haemophilus, Lactococcus

Sample names:
metaphlan_single2



julia&gt; metaphlan_profile(&quot;test/files/metaphlan_single2.tsv&quot;, :genus, sample = &quot;sample2&quot;)
CommunityProfile{Float64, Taxon, MicrobiomeSample} with 40 features in 1 samples

Feature names:
Prevotella, Roseburia, Faecalibacterium...Haemophilus, Lactococcus

Sample names:
sample2</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BiobakeryUtils.jl/blob/971791c639d54143c90ad2743182bfaf5917376a/src/metaphlan.jl#L82-L146">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.metaphlan_profiles" href="#BiobakeryUtils.metaphlan_profiles"><code>BiobakeryUtils.metaphlan_profiles</code></a> ‚Äî <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">metaphlan_profiles(paths::Array{&lt;:AbstractString, 1}, level::Union{Int, Symbol}=:all)</code></pre><p>Compiles MetaPhlAn profiles from multiple single tables into a CommunityProfile.</p><p>Examples ‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°</p><pre><code class="language-julia-repl hljs">julia&gt; metaphlan_profiles([&quot;test/files/metaphlan_single1.tsv&quot;, &quot;test/files/metaphlan_single2.tsv&quot;])
CommunityProfile{Float64, Taxon, MicrobiomeSample} with 129 features in 2 samples

Feature names:
Bacteria, Firmicutes, Bacteroidetes...Coprococcus_eutactus, Ruminococcus_bromii

Sample names:
metaphlan_single1, metaphlan_single2



julia&gt; metaphlan_profiles([&quot;test/files/metaphlan_single1.tsv&quot;, &quot;test/files/metaphlan_single2.tsv&quot;], :genus)
CommunityProfile{Float64, Taxon, MicrobiomeSample} with 46 features in 2 samples

Feature names:
Bacteroides, Roseburia, Faecalibacterium...Ruthenibacterium, Haemophilus

Sample names:
metaphlan_single1, metaphlan_single2



julia&gt; metaphlan_profiles([&quot;test/files/metaphlan_single1.tsv&quot;, &quot;test/files/metaphlan_single2.tsv&quot;], 5)
CommunityProfile{Float64, Taxon, MicrobiomeSample} with 24 features in 2 samples

Feature names:
Lachnospiraceae, Ruminococcaceae, Bacteroidaceae...Clostridiales_unclassified, Pasteurellaceae

Sample names:
metaphlan_single1, metaphlan_single2</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BiobakeryUtils.jl/blob/971791c639d54143c90ad2743182bfaf5917376a/src/metaphlan.jl#L264-L304">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.metaphlan_profiles" href="#BiobakeryUtils.metaphlan_profiles"><code>BiobakeryUtils.metaphlan_profiles</code></a> ‚Äî <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">metaphlan_profiles(path::AbstractString, level::Union{Int, Symbol}=:all; keepunidentified=false)</code></pre><p>Compiles MetaPhlAn profiles from a merged table into a CommunityProfile. Can select data according to taxonomic level. If level not given, all data is compiled. Set <code>keepunidentified</code> flag to <code>true</code> to keep <code>UNIDENTIFIED</code> data.</p><p>Levels may be given either as numbers or symbols:</p><ul><li><code>1</code> = <code>:kingdom</code></li><li><code>2</code> = <code>:phylum</code></li><li><code>3</code> = <code>:class</code></li><li><code>4</code> = <code>:order</code></li><li><code>5</code> = <code>:family</code></li><li><code>6</code> = <code>:genus</code></li><li><code>7</code> = <code>:species</code></li><li><code>8</code> = <code>:subspecies</code></li></ul><pre><code class="language-julia-repl hljs">
Examples
‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°
julia&gt; metaphlan_profiles(&quot;test/files/metaphlan_multi_test.tsv&quot;)
CommunityProfile{Float64, Taxon, MicrobiomeSample} with 42 features in 7 samples

Feature names:
Archaea, Euryarchaeota, Methanobacteria...Actinomyces_viscosus, GCF_000175315

Sample names:
sample1_taxonomic, sample2_taxonomic, sample3_taxonomic...sample6_taxonomic, sample7_taxonomic



julia&gt; metaphlan_profiles(&quot;test/files/metaphlan_multi_test.tsv&quot;, :genus)
CommunityProfile{Float64, Taxon, MicrobiomeSample} with 3 features in 7 samples

Feature names:
Methanobrevibacter, Methanosphaera, Actinomyces

Sample names:
sample1_taxonomic, sample2_taxonomic, sample3_taxonomic...sample6_taxonomic, sample7_taxonomic



julia&gt; metaphlan_profiles(&quot;test/files/metaphlan_multi_test.tsv&quot;, 3)
CommunityProfile{Float64, Taxon, MicrobiomeSample} with 2 features in 7 samples

Feature names:
Methanobacteria, Actinobacteria

Sample names:
sample1_taxonomic, sample2_taxonomic, sample3_taxonomic...sample6_taxonomic, sample7_taxonomic


julia&gt; metaphlan_profiles(&quot;test/files/metaphlan_multi_test_unidentified.tsv&quot;)
CommunityProfile{Float64, Taxon, MicrobiomeSample} with 43 features in 7 samples

Feature names:
UNIDENTIFIED, Archaea, Euryarchaeota...Actinomyces_viscosus, GCF_000175315

Sample names:
sample1_taxonomic, sample2_taxonomic, sample3_taxonomic...sample6_taxonomic, sample7_taxonomic



# julia&gt; metaphlan_profiles(&quot;test/files/metaphlan_multi_test_unidentified.tsv&quot;, keepunidentified = true)
# CommunityProfile{Float64, Taxon, MicrobiomeSample} with 43 features in 7 samples

# Feature names:
# UNIDENTIFIED, Archaea, Euryarchaeota...Actinomyces_viscosus, GCF_000175315

# Sample names:
# sample1_taxonomic, sample2_taxonomic, sample3_taxonomic...sample6_taxonomic, sample7_taxonomic</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BiobakeryUtils.jl/blob/971791c639d54143c90ad2743182bfaf5917376a/src/metaphlan.jl#L168-L242">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.parsetaxa-Tuple{AbstractString}" href="#BiobakeryUtils.parsetaxa-Tuple{AbstractString}"><code>BiobakeryUtils.parsetaxa</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">parsetaxa(taxstring::AbstractString; throw_error=true)</code></pre><p>Given a string representing taxonmic levels as formatted by MetaPhlAn (eg &quot;k<strong>Bacteria|p</strong>Proteobacteria...&quot;), separates taxonomic levels into elements of type Taxon in a vector.</p><p>Examples ‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°</p><pre><code class="language-julia-repl hljs">julia&gt; parsetaxa(&quot;k__Archaea|p__Euryarchaeota|c__Methanobacteria&quot;; throw_error = true)
3-element Vector{Taxon}:
 Taxon(&quot;Archaea&quot;, :kingdom)
 Taxon(&quot;Euryarchaeota&quot;, :phylum)
 Taxon(&quot;Methanobacteria&quot;, :class)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BiobakeryUtils.jl/blob/971791c639d54143c90ad2743182bfaf5917376a/src/metaphlan.jl#L363-L378">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="BiobakeryUtils.parsetaxon-Tuple{AbstractString}" href="#BiobakeryUtils.parsetaxon-Tuple{AbstractString}"><code>BiobakeryUtils.parsetaxon</code></a> ‚Äî <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">parsetaxon(taxstring::AbstractString, taxlevel::Union{Int, Symbol})</code></pre><p>Finds given taxonomic level in a string (as formatted by MetaPhlAn (eg &quot;k<strong>Bacteria|p</strong>Proteobacteria...&quot;)) and returns the clade and taxonomic level as a Taxon. If taxon level not given, function will return the most specific (lowest) taxonomic level available.</p><p>Levels may be given either as numbers or symbols:</p><ul><li><code>1</code> = <code>:kingdom</code></li><li><code>2</code> = <code>:phylum</code></li><li><code>3</code> = <code>:class</code></li><li><code>4</code> = <code>:order</code></li><li><code>5</code> = <code>:family</code></li><li><code>6</code> = <code>:genus</code></li><li><code>7</code> = <code>:species</code></li><li><code>8</code> = <code>:subspecies</code></li></ul><p>Examples ‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°‚â°</p><pre><code class="language-julia-repl hljs">julia&gt; parsetaxon(&quot;k__Archaea|p__Euryarchaeota|c__Methanobacteria&quot;, 2)
Taxon(&quot;Euryarchaeota&quot;, :phylum)

julia&gt; parsetaxon(&quot;k__Archaea|p__Euryarchaeota|c__Methanobacteria&quot;, :kingdom)
Taxon(&quot;Archaea&quot;, :kingdom)

julia&gt; parsetaxon(&quot;k__Archaea|p__Euryarchaeota|c__Methanobacteria&quot;)
Taxon(&quot;Methanobacteria&quot;, :class)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/BiobakeryUtils.jl/blob/971791c639d54143c90ad2743182bfaf5917376a/src/metaphlan.jl#L319-L349">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../humann/">¬´ Working with HUMAnN</a><a class="docs-footer-nextpage" href="../contributing/">Contributing ¬ª</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.7 on <span class="colophon-date" title="Wednesday 6 October 2021 20:13">Wednesday 6 October 2021</span>. Using Julia version 1.6.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
